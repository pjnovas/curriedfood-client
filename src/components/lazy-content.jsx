import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { Text } from 'react-native-paper';
import { useFocusEffect } from '@react-navigation/native';
import noop from 'lodash/noop';

import { composeHooks } from 'utils/language';
import Layout from './layout';
import Spinner from './spinner';

export const LazyContent = ({ showActivity, data, error, View }) => (
  <Layout style={{ paddingTop: showActivity ? 0 : 4 }}>
    {showActivity && <Spinner bar />}
    {error && <Text>Error: {JSON.stringify(error)}</Text>}
    {data && <View data={data} />}
  </Layout>
);

LazyContent.propTypes = {
  data: PropTypes.array,
  isLoading: PropTypes.bool,
  error: PropTypes.any,
  View: PropTypes.any,
  revalidate: PropTypes.func
};

LazyContent.defaultProps = {
  isValidating: false,
  isLoading: false,
  revalidate: noop
};

export const useFocus = ({ revalidate, isValidating, data, isLoading }) => {
  useFocusEffect(
    useCallback(() => {
      revalidate();
    }, [revalidate])
  );

  return {
    showActivity: !!(isValidating || data === null || isLoading)
  };
};

export default composeHooks({ useFocus })(LazyContent);
