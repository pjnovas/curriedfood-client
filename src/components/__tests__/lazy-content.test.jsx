import * as React from 'react';
import { View } from 'react-native';

import { useFocusEffect } from '@react-navigation/native';
import { LazyContent, useFocus } from '../lazy-content';

jest.mock('@react-navigation/native', () => ({
  useFocusEffect: jest.fn()
}));

takeSnapshots(
  {
    default: {},
    loading: {
      showActivity: true
    },
    withError: {
      isLoading: false,
      error: 'Boom!'
    },
    withData: {
      isLoading: false,
      View,
      data: ['a', 'b', 'c']
    }
  },
  LazyContent
);

describe('useFocus', () => {
  let spyUseCallback;
  beforeAll(() => {
    spyUseCallback = jest.spyOn(React, 'useCallback');
    React.useCallback.mockImplementation();
  });

  afterAll(() => {
    spyUseCallback.mockReset();
  });

  test('should return whenever to show spinner', () => {
    expect(useFocus({ data: [] })).toEqual({
      showActivity: false
    });

    expect(useFocus({ isValidating: true })).toEqual({
      showActivity: true
    });

    expect(useFocus({ data: null })).toEqual({
      showActivity: true
    });

    expect(useFocus({ isLoading: true })).toEqual({
      showActivity: true
    });

    expect(useFocusEffect).toBeCalled();
    expect(React.useCallback).toBeCalled();
  });
});
