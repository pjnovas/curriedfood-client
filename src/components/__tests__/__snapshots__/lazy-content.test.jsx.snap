// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
<Layout
  style={
    Object {
      "paddingTop": 4,
    }
  }
/>
`;

exports[`loading 1`] = `
<Layout
  style={
    Object {
      "paddingTop": 0,
    }
  }
>
  <Spinner
    bar={true}
  />
</Layout>
`;

exports[`withData 1`] = `
<Layout
  style={
    Object {
      "paddingTop": 4,
    }
  }
>
  <View
    data={
      Array [
        "a",
        "b",
        "c",
      ]
    }
  />
</Layout>
`;

exports[`withError 1`] = `
<Layout
  style={
    Object {
      "paddingTop": 4,
    }
  }
>
  <withTheme(Text)>
    Error: 
    "Boom!"
  </withTheme(Text)>
</Layout>
`;
